### ISSUE TITLE
Question: simultaneously train yolo for object detection and tracking using auxiliary layers and cosine similarity

### ISSUE THREAD

**Issue by pfeatherstone** [AlexeyAB/darknet#6004]
Can you fork some of the intermediate outputs of the backbone network into a few auxiliary layers which can be trained, using a cosine similarity loss function (or something similar), to output features suitable for tracking objects. Maybe this could be a two-stage training process, or this could be done end to end. 

I've been working on some object tracking using detections. Kalman filter solutions don't always work and it's the same story with optfow methods. The best i've seen so far, is to use cross correlation techniques (like the correlation tracker in the dlib library) but they are quite slow and scales with the number of detections. Methods like deepsort use a separate network to extract features which can be used to compare objects. But the yolo network is already calculating features. Surely, you only need to fork off some of those features, maybe pass them through a few additional conv layers, into a cosine similarity loss function, and voila. 

Am I crazy or does this sound sensible?

**Comment by AlexeyAB** [comment]
This is in progress: Conv-LSTM + Constrastive loss (with cosine similarity) + Yolo

**Comment by pfeatherstone** [comment]
Do you need the lstm bit ? Are you trying to do re-identification inside the network? I was thinking doing re-identification post processing using the feature set of each detected object with those of the previous frame. Essentially doing deepsort but recycling the features calculated by the detection network, and letting the hungarian algorithm do the re-identification by maximising similarity.

**Comment by pfeatherstone** [comment]
i was thinking the features could be added to the list of yolo features. So instead of having features
[tx,ty,tw,th,p,c0,...,c79], have [tx,ty,tw,th,p,c1,...,c79,f0,....,fN-1] at each yolo layer. The features aren't taken from the last conv layer the yolo layer saw, they are taken from earlier on.

**Comment by pfeatherstone** [comment]
actually that might only make sense for training end-to-end detection + tracking. I was hoping to just fork off some features, and use them as input to another 'tracking' network with a very minimal set of layers, which wouldn't require retraining the detection network.

**Comment by rbgreenway** [comment]
Use the same features to track that were used to detect.  This makes so much sense.  Very excited that this is being worked on.  Alexy, thanks for your excellent work and dedication to this technology.

**Comment by ntd94** [comment]
I personally think that a detection network tries to generalize the characteristics of a class, as all people have to have similar features. On the other hand, tracking requires contrastiveness, a person's feature needs to be as far as possible from the rest of the people. Training a network serving conflict targets might be unrealistic. That's why we haven't found such a network like that.

**Comment by pfeatherstone** [comment]
Yeah. I think that's why you might need to attach an auxiliary network that uses some of the features output by the early inner layers of the detection network, that is trained separately using similarity loss or contrastive loss. Basically you end up with deepsort, but instead of using a completely separate VGG16 network, you're using X% of the detection network and (100-X)% of a new auxiliary network. It feels slightly more efficient and faster to recycle features. I like deepsort because all the re-id is undertaken by the hungarian algorithm and all you need is a good metric, and in this case, some good tracking features to evaluate the metric on. The only problem with deepsort is that it is super slow. So recyling features from a detection network seems like a good way to solve that. But it would be interesting to see if @AlexeyAB's idea of a fully blown LTSM is better and more accurate.

**Comment by AlexeyAB** [comment]
Yes, these 2 tasks (Detection and Reidentification) partially contradict each other. Therefore, is expected: a slight decrease in accuracy in images, but a large increase in accuracy in video.
I think we can regulate this through the normalization factor.


But perhaps the task of re-identification will make network remember more details of objects, that theoretically can even improve detection even in images for large networks (networks with high capacity).

**Comment by pfeatherstone** [comment]
you make a good point. Maybe re-id using LSTM will make the detection network pay special attention to object features. If you're right, then maybe you need to write a paper with title "LSTM is all the attention you need" :)

**Comment by AlexeyAB** [comment]
> Conv-LSTM + Constrastive loss (Embeddings for cosine similarity) + Yolo

This is done, just we should improve it and test it.

**Comment by pfeatherstone** [comment]
Awesome! Have you posted some weights?

**Comment by AlexeyAB** [comment]
Not yet.

**Comment by AlexeyAB** [comment]
There is the Proof of concept cfg-file, you can try to train `Constrastive loss (Embeddings for cosine similarity) + Yolo` (without Conv-LSTM) on any non-sequence-datasest (MSCOCO, BDD, OpenImages, PascalVOC, ...): [yolov3-tiny_contrastive.cfg.txt](https://github.com/AlexeyAB/darknet/files/4887210/yolov3-tiny_contrastive.cfg.txt)

Train as usual, without pre-trained weights or with https://drive.google.com/file/d/18v36esoXCh-PsOKwyP2GWrpYDptDY8Zf/view?usp=sharing

This model will count your objects, when you will run it on Video 
`./darknet detector demo data/sobj.data yolov3-tiny_contrastive.cfg backup/yolov3-tiny_contrastive_last.weights video.avi -out_filename out_video.avi`

You can play with parameters for Detection (after training):
```
[yolo]
# for tracking
track_history_size = 5 - find similiraty on 5 previous frames
sim_thresh = 0.8 - similarity threshold to consider an object on two frames the same
dets_for_show = 2 - number of frames with this object before Show it
dets_for_track = 8 - number of frames with this object before Track it
track_ciou_norm = 0.3 - take into account CIoU (0.0 to 1.0)
```

**Comment by pfeatherstone** [comment]
Thanks @AlexeyAB

**Comment by AlexeyAB** [comment]
I re-upload contrastive detection model [yolov3-tiny_contrastive.cfg.txt](https://github.com/AlexeyAB/darknet/files/4887786/yolov3-tiny_contrastive.cfg.txt) with [contrastive] cls_normalizer=1.0  - use this
(previously I uploaded [contrastive] cls_normalizer=0.0 – so contrastive loss was disabled)

**Comment by AlexeyAB** [comment]
Simple test on small dataset:

* Contrastive loss doesn't degrade mAP of Detector
* Contrastive loss improves accuracy of Tracking (so we can build DeepSort on top of Yolo with much higher accuracy than without Contrastive loss)

| Contrastive loss enabled | Contrastive loss disabled |
|---|---|
| ![det_cl_fwonly_mi_b64](https://user-images.githubusercontent.com/4096485/86916966-0af8fb80-c12d-11ea-85ba-05d01b3739e9.png) | ![det_cl_fwonly_mi_b64_cl-disabled](https://user-images.githubusercontent.com/4096485/86916983-10564600-c12d-11ea-9b36-6f3850ce3fcd.png) |

**Comment by pfeatherstone** [comment]
This is good news

**Comment by MsWik** [comment]
What could be the problem?

 Warning: in txt-labels class_id=-2147483648 >= classes=1 in cfg-file. In txt-labels class_id should be [from 0 to 0] 

 truth.x = 0.000000, truth.y = -nan, truth.w = 0.000000, truth.h = 0.000000, class_id = -2147483648

Wrong label: truth.x = -0.000000, truth.y = 0.000000, truth.w = -0.000000, truth.h = 0.000000 
 Wrong label: truth.x = -nan, truth.y = -885731875545466011648.000000, truth.w = -0.000000, truth.h = 0.000000 
 Wrong label: truth.x = 0.000000, truth.y = 0.000000, truth.w = -0.000000, truth.h = 0.000000 
v3 (mse loss, Normalizer: (iou: 0.75, cls: 1.00) Region 38 Avg (IOU: 0.000000, GIOU: 0.000000), Class: 0.000000, Obj: 0.000000, No Obj: 0.000000, .5R: 0.000000, .75R: 0.000000, count: 3, class_loss = 0.750000, iou_loss = -nan, total_loss = -nan 
 Contrast accuracy = 0.000000 % 
 Error: N == 0 || temperature == 0 || vec_len == 0. N=1.000000, temperature=nan, vec_len=0.000000 

Standard Labeling, 1 class (0)

0 0.241 0.854 0.149 0.285
0 0.518 0.529 0.126 0.24

With other cfg there is no problem.
--

**Comment by AlexeyAB** [comment]
Look at bad.list and bad_label.list files.

Try to download the latest Darknet version, and recompile. I added minor fix.

What dataset do you use and how many images?

**Comment by MsWik** [comment]
Thank. Yes, that helped. I need to find and track cars. I use COCO-17 + my data (~ 15000 photos).

**Comment by AlexeyAB** [comment]
Yes, it will work. But it will work much better when yolov4-tiny-contrastive.cfg will be implemented.
Currently there is used very simple and inaccurate model: 
https://drive.google.com/file/d/1g18BgkIRbZGykHYxKvWgH1_UUTQnONwG/view?usp=sharing

**Comment by pfeatherstone** [comment]
Are you training these models on MOT datasets? Not quite sure how the contrastive loss works on COCO if no two objects are related. Doesn't that mean you only ever have negative samples. Do you need tracked objects to get positive samples? This is probably a stupid question.

**Comment by AlexeyAB** [comment]
* Contrastive loss gets at least 1 positive by using **2 copies of the same images which are augmented differently**

* Contrastive loss can use many positives and negatives unlike Triplet loss

**Comment by pfeatherstone** [comment]
I see, so the key here is augmentation

**Comment by pfeatherstone** [comment]
> Simple test on small dataset:
> 
>     * Contrastive loss doesn't degrade mAP of Detector
> 
>     * Contrastive loss improves accuracy of Tracking (so we can build DeepSort on top of Yolo with much higher accuracy than without Contrastive loss)
> 
> 
> Contrastive loss is enabled `[contrastive] cls_normalizer=1.0` 	Contrastive loss is disabled `[contrastive] cls_normalizer=0.0` 	Contrastive, flip and jitter are enabled `[net] contrastive_jit_flip=1`
> ![det_cl_fwonly_mi_b64](https://user-images.githubusercontent.com/4096485/86916966-0af8fb80-c12d-11ea-85ba-05d01b3739e9.png) 	![det_cl_fwonly_mi_b64_cl-disabled](https://user-images.githubusercontent.com/4096485/86916983-10564600-c12d-11ea-9b36-6f3850ce3fcd.png) 	![det_cl_fwonly_mi_b64_jitter](https://user-images.githubusercontent.com/4096485/87153627-71fae980-c2c0-11ea-900b-9d3f6f4ec01a.png)

If the constrastive loss depends highly on augmentation, should there be other augmentation transformations like stretching, warping, etc?

**Comment by AlexeyAB** [comment]
@pfeatherstone 

It depends on how much the same object can differ in two frames:

* If we have video with 25-60 fps for tracking slow moving persons, then the same person will not very different on two frames - we don't need to use strong data augmentation

* If we have video with 1-5 FPS for tracking fast moving cars / birds, or we want to track one object on two cameras, then the same car can be very different on two frames/cameras - we need to use strong data augmentation

----

Also it depends on model:

* Contrastive-learning require strong data augmentation
* Tiny-model require weak data augmentation for good AP

So we should use strong data augmentation only for big `yolov4-contrastive` model rather than for `yolov4-tiny-contrastive`.

**Comment by MsWik** [comment]
Hey. Tell me how soon you will publish yolov4-contrastive.cfg. Or maybe there is a test model for testing. Thank.

**Comment by MsWik** [comment]
Hello. Tell me if there will be enough 1 [contrastive] layer in yolov4 / yolov4-tiny or need more? We write parameter tracking in only one yolo layer ?
Thanks for your work

**Comment by AlexeyAB** [comment]
There should be 1 [contrastive] layer and 1 - 5 [yolo] layers.

It will be tested. For  simplicity of implementation and calculation of embeddings, [contrastive] layer can work only with 1 [yolo] layer so far.

**Comment by MsWik** [comment]
Hello. Thanks for the answer. I got a good graph but no object number. I am not quite sure what the "embedding_layer" should be pointing to. Attached is the cfg and training schedule.

[yolov4-tiny-3l-supertest.cfg.txt](https://github.com/AlexeyAB/darknet/files/4930225/yolov4-tiny-3l-supertest.cfg.txt)
[result](https://user-images.githubusercontent.com/10869230/87646132-11076180-c757-11ea-8fa5-19db5dfc67d4.png)

**Comment by alexanderfrey** [comment]
Do you have an example on how to extract the embeddings for a given frame ?

**Comment by AlexeyAB** [comment]
> Hello. Thanks for the answer. I got a good graph but no object number. I am not quite sure what the "embedding_layer" should be pointing to. Attached is the cfg and training schedule.
> 
> yolov4-tiny-3l-supertest.cfg.txt
> result


* embedding_layer" should be pointing to [convolutional] layer before [contrastive] layer as in the example: https://github.com/AlexeyAB/darknet/files/4887786/yolov3-tiny_contrastive.cfg.txt

* Currently [contrastive] layer supports only 1 [yolo] layer, so you should use only 1 [yolo] layer in your cfg file isntead of 3

**Comment by AlexeyAB** [comment]
> Do you have an example on how to extract the embeddings for a given frame ?

If you use DLL/SO library:

C API: https://github.com/AlexeyAB/darknet/blob/20760d29715bb34bb2fbd0a05318dafe8150b325/include/darknet.h#L864-L878

Python example: https://github.com/AlexeyAB/darknet/blob/20760d29715bb34bb2fbd0a05318dafe8150b325/darknet.py#L56-L68



Using `./darknet detector demo ...` you will see `track_id`

**Comment by MsWik** [comment]
Thanks for the answer. I will wait. The question is in what cases the error "number of names 1 that isn't equal to classes = 0 in the file" occurs just in the cfg file classes = 1. Thank

**Comment by MsWik** [comment]
Layers [contrasting] should be exactly in such? Why can't we do it like this? (this solution gives a good AP) 
[yolov4-nn.cfg.txt](https://github.com/AlexeyAB/darknet/files/4957890/yolov4-nn.cfg.txt)
![nn](https://user-images.githubusercontent.com/10869230/88135602-26194000-cbf0-11ea-98ed-44edc4dc9204.png)

**Comment by MsWik** [comment]
Labels are displayed but not correct. I tried to change sim_thresh. What could be the reason? Thanks in advance for your reply.
[yolov4-nn-2207n.cfg.txt](https://github.com/AlexeyAB/darknet/files/4972091/yolov4-nn-2207n.cfg.txt)

**Comment by AlexeyAB** [comment]
Your cfg-file is incorrect.

`[yolo] embedding_layer = 8` can't point to one of the next layers. It should point to the one of the previous layers.

**Comment by Anafeyka** [comment]
> Your cfg-file is incorrect.
> 
> `[yolo] embedding_layer = 8` can't point to one of the next layers. It should point to the one of the previous layers.

if 
`[contrastive]
classes=1
temperature=1.0
yolo_layer= -8
cls_normalizer=1.0
max_delta=10`

then
`embedding_layer = -4`
Did I understand correctly?
[testTiny4Contr.txt](https://github.com/AlexeyAB/darknet/files/4973834/testTiny4Contr.txt)

**Comment by AlexeyAB** [comment]
[yolo] embedding_layer = -4 should point to the conv-layer which precedes the [contrasting] layer.

Don't change last 7 layers: https://github.com/AlexeyAB/darknet/files/4887782/yolov3-tiny_contrastive.cfg.txt
You can only change other layers, and you can change parameters in [yolo] layer to make it yolov4-tiny.

**Comment by Anafeyka** [comment]
> [yolo] embedding_layer = -4 should point to the conv-layer which precedes the [contrasting] layer.
> 
> Don't change last 7 layers: https://github.com/AlexeyAB/darknet/files/4887782/yolov3-tiny_contrastive.cfg.txt
> You can only change other layers, and you can change parameters in [yolo] layer to make it yolov4-tiny.

Like this. Or do I need to sleep
[testTiny4Contr2.txt](https://github.com/AlexeyAB/darknet/files/4973958/testTiny4Contr2.txt)

**Comment by AlexeyAB** [comment]
@Anafeyka Yes!

**Comment by Anafeyka** [comment]
> @Anafeyka Yes!

thank! Now I understand how it works

**Comment by rbgreenway** [comment]
Is there an example cfg file for full YoloV4 network that includes the embedding?  I'd really like to train a network and test the tracking based on the calculated embedding.

**Comment by MsWik** [comment]
Hey. Tell me how to get similarity identifiers through the py script? Can I use cv 4.4 / cuDNN with the [contrastive] layer? Thank.

**Comment by zpmmehrdad** [comment]
@AlexeyAB Hi,

How can I use tracking? Could you please give us an example?

Thanks

**Comment by haviduck** [comment]
trying to understand @Anafeyka cfg, and wanting to try with 2 classes, what would the convolutional filters be at the end?

**Comment by Deqm525** [comment]
@MsWik Hi, 

Have you successfully trained full yolov4+contrastive? Could you please share the cfg? My version always throw me random segmentation faults

Thanks

**Comment by scianand** [comment]
Hi Please can you provide me cfg of yolov3-tiny_contrastive model cause I am having segmentation fault errors?

**Comment by scianand** [comment]
Please someone can tell me what's the reason behind this segmentation fault error?

**Comment by mrxuehb** [comment]
I have also encountered segmentation fault when using yolov3-tiny_contrastive.cfg. I set classes = 7, and filters=108. It can be trained for 200-600 epochs randomly before failing with segmentation error.  Is this  yolov3-tiny_contrastive.cfg can only used with classes = 1 currently? Or is there other problems?

**Comment by MsWik** [comment]
Sori, I haven't watched the topic for a long time. If relevant - here's my cfg.
[yolov4-nn-22_07.cfg.txt](https://github.com/AlexeyAB/darknet/files/5232390/yolov4-nn-22_07.cfg.txt)

**Comment by Mahdi89** [comment]
@AlexeyAB in your cfg file what's the relationship between `classes=1` `filters=54` and `num=9`?  asking to see how to set them for a dataset for 2 classes. Thanks

**Comment by javier-box** [comment]
Hello @MsWik ... Did you get contrastive to work? Can you share your test result?

I'm also wondering how to add contrastive layer to tkdnn / cudnn.

**Comment by AlexeyAB** [comment]
Experimental contrastive model that is trained on MSCOCO:

* cfg: https://raw.githubusercontent.com/AlexeyAB/darknet/master/cfg/yolov4-tiny_contrastive.cfg

* weights: https://github.com/AlexeyAB/darknet/releases/download/darknet_yolo_v4_pre/yolov4-tiny_contrastive_last.weights


`./darknet detector demo cfg/coco.data cfg/yolov4-tiny_contrastive.cfg yolov4-tiny_contrastive_last.weights test.avi -out_filename out_test.avi -ext_output`

**Comment by haviduck** [comment]
thats really cool. Great fps and solid tracking at This stage. frozen offscreen dets  were picked up on reentry. Coco Will as always tell me a scooter is a flower pot but thats not the mot's fault. This is awesome,  thanks for sharing!

**Comment by AlexeyAB** [comment]
I added some hyper-parameters:
* for training: https://github.com/AlexeyAB/darknet/blob/8297b1be3108f47d82b97d322d5f1aef5150018f/cfg/yolov4-tiny_contrastive.cfg#L18-L20
* for tracking: https://github.com/AlexeyAB/darknet/blob/8297b1be3108f47d82b97d322d5f1aef5150018f/cfg/yolov4-tiny_contrastive.cfg#L437-L441

This is just a proof of concept, so tracking is very poor.

**Comment by Hwijune** [comment]
> I added some hyper-parameters:
> 
> * for training: https://github.com/AlexeyAB/darknet/blob/8297b1be3108f47d82b97d322d5f1aef5150018f/cfg/yolov4-tiny_contrastive.cfg#L18-L20
> * for tracking: https://github.com/AlexeyAB/darknet/blob/8297b1be3108f47d82b97d322d5f1aef5150018f/cfg/yolov4-tiny_contrastive.cfg#L437-L441
> 
> This is just a proof of concept, so tracking is very poor.

Hi! @AlexeyAB 

How do I apply tracking to test?

`./darknet` detector demo ./cfg/coco.data ./yolov4-tiny_contrastive.cfg ./yolov4-tiny_contrastive_last.weights Human_Dog_Horse.mp4 -i 0 -thresh 0.25

**Comment by Hwijune** [comment]
Hi! @AlexeyAB 

![image](https://user-images.githubusercontent.com/37058972/96812942-4ce94b80-1457-11eb-925e-741eac65a10a.png)

If we have video with 1-5 FPS for tracking fast moving cars / birds, or we want to track one object on two cameras, then the same car can be very different on two frames/cameras - we need to use strong data augmentation

training
pic1 origin image 
pic2 using augment image 

Does the higher the Contrast loss, the higher the tracking accuracy in the video?

**Comment by syjeon121** [comment]
hi, @AlexeyAB 
is this feature (tracking using contrastive loss) available in c++ api, yolo_v2_class.hpp ??

**Comment by zpmmehrdad** [comment]
@AlexeyAB Hi,

Can we use the tracking in Open-CV?

Thanks

**Comment by arnaud-nt2i** [comment]
@zpmmehrdad Yes you can, it's called Optiflow and it works quite well.
Lot's of other topics about this ( you have to compile OpenCV with Cuda support (from source))

**Comment by Goru1890** [comment]
How do I can use tracking with Yolov4-contrastive?
Is the tracker already available in darknet?
I'm interested in a multi-object tracker so I can even count for how much time an object has been tracked until it disappears from the screen. Using as input a video.

**Comment by Goru1890** [comment]
How does contrastive work with the tracker?

So contrastive compares to object in two different frames and then it decides if it is the same object and then track it?

**Comment by Goru1890** [comment]
Do you have a pre-trained model for yolov4-tiny-contrastive?

**Comment by MuhammadAsadJaved** [comment]
@Goru1890  The model and .cfg given in this same issue. Check a few last comments.

**Comment by AdamCuellar** [comment]
For anyone looking to train the contrastive model and avoid a segfault, I've been able to train for 7k+ iterations with no issues after initializing the variables in the for loops in softmax_layer.c starting on line 502. The "#pragma omp parallel for" caused the variables to go out of bounds, making the z_index larger than z. Specifically, I defined int b3, n3, h3, and w3 in the respective for statement then changed the corresponding variables in the loop. 

https://github.com/AlexeyAB/darknet/blob/e2a128737bcee224088a0076629983e2a4beca01/src/softmax_layer.c#L502-L511


@AlexeyAB could you point me in the right direction for implementing the contrastive layer for more than one yolo head?

**Comment by slidespike** [comment]
I'm trying to train on my own data set and can't pass the first round, (without any specific error it's just stopped) can someone help me please?

I change the number of classes to 4 and calculated the number of filter in the upper layer to be (4+5)x9 = 81

do I need to change something else?

**Comment by MsWik** [comment]
https://yadi.sk/i/WpZQhZYtUd7A4ghttps://github.com/AlexeyAB/darknet#how-to-train-with-multi-gpu https://disk.yandex.by/i/kWlOMt9r8xEYMA  12.05.2021, 12:17, "slidespike" ***@***.***>:
I'm trying to train on my own data set and can't pass the first round, (without any specific error it's just stopped) can someone help me please?
I change the number of classes to 4 and calculated the number of filter in the upper layer to be (4+5)x9 = 81
do I need to change something else?

**Comment by lukastruemper** [comment]
> For anyone looking to train the contrastive model and avoid a segfault, I've been able to train for 7k+ iterations with no issues after initializing the variables in the for loops in softmax_layer.c starting on line 502. The "#pragma omp parallel for" caused the variables to go out of bounds, making the z_index larger than z. Specifically, I defined int b3, n3, h3, and w3 in the respective for statement then changed the corresponding variables in the loop.
> 
> https://github.com/AlexeyAB/darknet/blob/e2a128737bcee224088a0076629983e2a4beca01/src/softmax_layer.c#L502-L511
> 
> @AlexeyAB could you point me in the right direction for implementing the contrastive layer for more than one yolo head?

@AdamCuellar 

I can confirm that this solves the segfault problem, but I couldn't reconstruct your solution from the description. Removed the omp pragma completely for now and it works without problems. Could you please share the code or even create a PR for that?

**Comment by AdamCuellar** [comment]
@lukastruemper 

Below is a screenshot of the changes I made to fix the segfault. If you have any other issues let me know!

![image](https://user-images.githubusercontent.com/40183888/127147557-6ce80a4f-6942-4b73-a8f7-e6101a55acf7.png)

**Comment by AlexeyAB** [comment]
@AdamCuellar So if it solves the issue (training goes well, and trained model detects correctly) you can create a Pull request for these changes.

**Comment by rbgreenway** [comment]
@AdamCuellar -  have you had a chance to see how well the tracking performs?

**Comment by AdamCuellar** [comment]
> 
> 
> @AdamCuellar So if it solves the issue (training goes well, and trained model detects correctly) you can create a Pull request for these changes.

@AlexeyAB I haven't had a chance to train it on a good dataset. Should I train on COCO to verify it works well enough? If so, I'll train and report the results + make a pull request.

@rbgreenway I trained on a small private dataset and the tracking performed well but due to the limited data the training/testing set were similar so I don't want to make any definitive statement. If you, or anyone, knows of a good dataset to test this on then I can train and post results.

**Comment by AlexeyAB** [comment]
@AdamCuellar 

I added a fix: https://github.com/AlexeyAB/darknet/commit/f5007cd0d40cf6207aceda564458d3fa021872b3
Please, try to test it by your small dataset or large COCO-dataset.

I think we should define `for`-loop-variables (`nd, hd, wd`) inside the `for`-statement for the inner loops, and outside for the OpenMP-`for`-loop variable `nb` https://docs.microsoft.com/en-us/cpp/parallel/openmp/reference/openmp-directives?view=msvc-160#for-openmp

**Comment by AdamCuellar** [comment]
@AlexeyAB 

Yes I agree, that's much better. I've started training on COCO. I will also start training later today on the small driving dataset you've used for the YOLO w/ LSTM so I can report back with some results.

**Comment by rbgreenway** [comment]
@AdamCuellar Thanks, Adam.  I'm very interested in how the tracking performs.  I won't have time for a couple weeks, but I'll certainly train it on a dataset that I have access to...particularly if you report some good results.  Thanks again.

**Comment by AdamCuellar** [comment]
@AlexeyAB Sadly, both experiments ended with a segfault. On the small driving set it got to almost 14k iterations and on COCO it got to about 5k iterations. 

I think declaring the iterator variable inside the parallelized for-loop might be necessary when there are nested for-loops. I will modify the code and try again to see if it still segfaults.

**Comment by AdamCuellar** [comment]
@AlexeyAB Looks like the changes you made work. For the small driving dataset you've used before, I trained for 20k iterations with no issues. Here is the loss chart:

![image](https://user-images.githubusercontent.com/40183888/127508410-203afdc6-936e-448b-989c-e80b549f77f9.png)

On COCO, it's gotten to 17k+ iterations with no issues. It will probably take a really long time though as I only had 1 gpu available to train.

**Comment by akashAD98** [comment]
@AlexeyAB @AdamCuellar can we use this same concept for different models such as yolov4x-mish,yolov4-csp,swish?
I want to do both task, .first it will detect objects  & even we can add tracker on the top of detected object.

i saw there are only two cfg availables yolov3,yolov4 contrasive for object tracking . im looking for mish,csp,swish -model
i dont know if its possible to do tracking on this model, Please let us know.  Thank you so much .

**Comment by AdamCuellar** [comment]
> @AlexeyAB @AdamCuellar can we use this same concept for different models such as yolov4x-mish,yolov4-csp,swish?
> I want to do both task, .first it will detect objects & even we can add tracker on the top of detected object.

@akashAD98 You can technically use the contrastive layer for any model, you just need to change the model to have 1 yolo head  and use the features prior to the yolo head as input to the contrastive layer. If you read through this thread, you'll get a better understanding of how to do this. 

> we can train object detection model & if we want to do object tracking we can simply change parameters inside cfg , we can add both parameters like:
> object_detection=1 "it will do only object detection"
> object_tracking=1 "it will do object tracking"
> 
> (this is just my thought process.it would be easier to use.)

I can see how this may seem like it would be easier to use; however, I think it would make things more obscure. An object_detection parameter is a bit redundant since the yolo networks were designed for object detection. An object_tracking parameter wouldn't provide enough information for tracking such as which layer is used for the contrastive loss, the number of detections for tracking, track history size, etc.

**Comment by akashAD98** [comment]
@AdamCuellar for training custom model should we need to use the same command? & few changes in cfg right?
`!./darknet detector train cfg/coco.data  cfg/yolov4-tiny_contrastive.cfg  yolov4-tiny_contrastive_last.weights`

& it will apply tracker on all classes which is present in the dataset?
